indico/modules/events/sessions/util.py

```diff
@@ -143,3 +146,31 @@ def session_coordinator_priv_enabled(event, priv):
     """
     from indico.modules.events.sessions import COORDINATOR_PRIV_SETTINGS, session_settings
     return session_settings.get(event, COORDINATOR_PRIV_SETTINGS[priv])
+
+
+def get_events_with_user_linked_sessions(user, from_dt=None, to_dt=None):
+    """Returns a dict with keys representing event_id and the values containing
+    data about the user rights for sessions within the event
+
+    :param user: A `User`
+    :param from_dt: The earliest event start time to look for
+    :param to_dt: The latest event start time to look for
+    :return: A dict with relevant data
+    """
+    event_date_filter = None
+    if from_dt and to_dt:
+        event_date_filter = IndexedEvent.start_date.between(from_dt, to_dt)
+    elif from_dt:
+        event_date_filter = IndexedEvent.start_date >= from_dt
+    elif to_dt:
+        event_date_filter = IndexedEvent.start_date <= to_dt
+
+    query = (user.in_session_acls
+             .options(load_only('session_id'))
+             .options(joinedload(SessionPrincipal.session))
+             .join(Session)
+             .join(Event, Event.id == Session.event_id)
+             .join(IndexedEvent, IndexedEvent.id == Session.event_id)
+             .filter(~Session.is_deleted, ~Event.is_deleted)
+             .filter(event_date_filter))
+    return {principal.session.event_id: principal.current_data for principal in query}
```

Comment:
Why not return a set of "roles" (i.e. the link types) from here, then you can just update `|=` the set in the api code